src/Nethermind/Nethermind.Blockchain/BlockhashCache.cs

```diff
@@ -0,0 +1,249 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Nethermind.Blockchain.Headers;
+using Nethermind.Core;
+using Nethermind.Core.Caching;
+using Nethermind.Core.Collections;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Threading;
+using Nethermind.Logging;
+using Nethermind.Specs;
+
+namespace Nethermind.Blockchain;
+
+public class BlockhashCache(IHeaderFinder headerFinder, ILogManager logManager) : IDisposable, IBlockhashCache
+{
+    private readonly ILogger _logger = logManager.GetClassLogger();
+    private readonly ConcurrentDictionary<Hash256AsKey, CacheNode> _blocks = new();
+    private readonly LruCache<Hash256AsKey, Hash256[]> _flatCache = new(32, nameof(BlockhashCache));
+    public const int MaxDepth = BlockhashProvider.MaxDepth;
+    private long _minBlock = int.MaxValue;
+    private Task _pruningTask = Task.CompletedTask;
+
+    public Hash256? GetHash(BlockHeader headBlock, int depth) =>
+        depth == 0 ? headBlock.Hash
+        : depth == 1 ? headBlock.ParentHash
+        : depth > MaxDepth ? null
+        : _flatCache.TryGet(headBlock.Hash!, out Hash256[] array) ? array[depth]
```

Comment:
You can query based on ParentHash instead of Hash, since it was populated from the last block, there is no need for the timing of the current Prefetch to be finished in this block.